tensorboard/manager_test.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Unit tests for `tensorboard.manager`."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import datetime
+
+import json
+import re
+import six
+import tensorflow as tf
+
+from tensorboard import manager
+from tensorboard import version
+
+
+def _make_info(i=0):
+  """Make a sample TensorboardInfo object.
+
+  Args:
+    i: Seed; vary this value to produce slightly different outputs.
+
+  Returns:
+    A type-correct `TensorboardInfo` object.
+  """
+  return manager.TensorboardInfo(
+      version=version.VERSION,
+      start_time=datetime.datetime.fromtimestamp(1548973541 + i),
+      port=6060 + i,
+      pid=76540 + i,
+      path_prefix="/foo",
+      logdir="~/my_data/",
+      db="",
+      cache_key="asdf",
+  )
+
+
+class TensorboardInfoTest(tf.test.TestCase):
+  """Unit tests for TensorboardInfo typechecking and serialization."""
+
+  def test_roundtrip_serialization(self):
+    # This will also be tested indirectly as part of `manager`
+    # integration tests.
+    info = _make_info()
+    also_info = manager._info_from_string(manager._info_to_string(info))
+    self.assertEqual(also_info, info)
+
+  def test_serialization_rejects_bad_types(self):
+    info = _make_info()._replace(start_time=1549061116)  # not a datetime
+    with six.assertRaisesRegex(
+        self,
+        ValueError,
+        "expected 'start_time' of type.*datetime.*, but found: 1549061116",
+    ):
```

Comment:
I will admit I see no style rule prohibiting this, but it is inconsistent with the rest of TensorBoard to separate out the closing paren like this when it's opening an indented block.  My eye reads the `with` and looks for the closing out-indent and finds this parent and gets very confused.

So if you're not particularly opposed, maybe let's keep the closeparen and colon on the end of the preceding line?  Unlike for e.g. a tuple definition where it's nice to be able to end each element with a comma for consistency when adding/removing them, in this case there will always be 3 arguments to this function so that seems less relevant.

Reply:
k